### PR TITLE
filefetcher not fetching Readme.md files despite uv/hatchling requiring it for metadata

### PR DESCRIPTION
**Pull request by a-schur**
fetches Readme files when it is required by uv

### What are you trying to accomplish?

Fetch README declared (or implicitly expected) by PEP 621 pyproject.toml during uv updates as an optional support file; prevents hatchling OSError when readme = "README.md" is present. Fallback probes common README names only if no explicit readme key. Specs updated to ignore support files to remove brittle exact-count assumptions.

Stops uv lock failures like: OSError: Readme file does not exist: [README.md](vscode-file://vscode-app/private/var/folders/q7/_8nm4mls1jb_y9h5vsts50yr0000gn/T/AppTranslocation/92DE8009-94D4-4A6C-9244-3AE1C018D68E/d/Visual%20Studio%20Code.app/Contents/Resources/app/out/vs/code/electron-browser/workbench/workbench.html) (hatchling metadata phase). Improves robustness for projects relying on README for package metadata.

New constant: README_FILENAMES (4 variants, bounded network impact).
README files marked support_file (no change to dependency parsing).
Per-candidate 404 rescued silently; pattern matches other ecosystems’ support fetchers.
Tests now filter out support files instead of asserting raw counts.
Consider if any other Python ecosystems should share this logic (future extraction).

Smallest change: augment existing fetch flow without introducing config flags; avoids premature optimization or broader pyproject parsing changes; keeps behavior additive and safe.

Dry run on an affected repo with readme = "README.md" no longer errors; uv lock completes.
All existing + adjusted specs pass.
Returned file set includes README only when present (or found via fallback) and flagged support_file.
No regression for repos without README (only up to 4 extra 404s, all rescued).

### Checklist

- [x] I have run the complete test suite to ensure all tests and linters pass.
- [x] I have thoroughly tested my code changes to ensure they work as expected, including adding additional tests for new functionality.
- [x] I have written clear and descriptive commit messages.
- [x] I have provided a detailed description of the changes in the pull request, including the problem it addresses, how it fixes the problem, and any relevant details about the implementation.
- [x] I have ensured that the code is well-documented and easy to understand.

### CODE REVIEW

File: uv/spec/dependabot/uv/file_fetcher_spec.rb [hunk 1]
```diff
@@ -745,11 +750,17 @@
         stub_request(:get, url + "setup.py?ref=sha")
           .with(headers: { "Authorization" => "token token" })
           .to_return(status: 404)
+        # README support files are probed; ensure they 404 so they're ignored without VCR noise
```
Comment (robaiken) [inline uv/spec/dependabot/uv/file_fetcher_spec.rb]:
```suggestion
```

File: uv/spec/dependabot/uv/file_fetcher_spec.rb [hunk 2]
```diff
@@ -745,11 +750,16 @@
...
+            .with(headers: { "Authorization" => "token token" })
+            .to_return(status: 404)
+        end
       end
 
       it "doesn't raise a path dependency error" do
-        expect(file_fetcher_instance.files.count).to eq(2)
-        expect(file_fetcher_instance.files.map(&:name)).to contain_exactly("requirements-test.txt", "pyproject.toml")
+        primary_files = file_fetcher_instance.files.reject(&:support_file?)
+        expect(primary_files.map(&:name)).to contain_exactly("requirements-test.txt", "pyproject.toml")
```
Comment (v-thavaahariharangit) [inline uv/spec/dependabot/uv/file_fetcher_spec.rb:905]: There is no test for actual purpose of this PR.

```
context "when pyproject.toml declares a README file" do
  before do
    stub_request(:get, url + "pyproject.toml?ref=sha")
      .to_return(status: 200, body: fixture_with_readme_declaration)
    
    # Actually test successful README fetching
    stub_request(:get, url + "README.md?ref=sha")
      .to_return(status: 200, body: '{"content": "base64content", "name": "README.md"}')
  end

  it "fetches the declared README as a support file" do
    files = file_fetcher_instance.files
    readme_files = files.select(&:support_file?)
    
    expect(readme_files.map(&:name)).to include("README.md")
    expect(files.find { |f| f.name == "README.md" }).to be_support_file
  end
end
```

File: uv/lib/dependabot/uv/file_fetcher.rb [hunk 3]
```diff
@@ -113,6 +118,41 @@ def pyproject_files
         [pyproject].compact
       end
 
+      sig { returns(T::Array[Dependabot::DependencyFile]) }
+      def readme_files
+        return [] unless pyproject
+
+        # Attempt to read the readme declaration from the pyproject. Accept both simplified
```
Comment (v-thavaahariharangit) [inline uv/lib/dependabot/uv/file_fetcher.rb:125]: Nice to have Document

```
# Fetches README files that may be referenced in pyproject.toml metadata.
#
# This is UV-specific functionality required because:
# 1. uv and hatchling read README content during dependency resolution
# 2. Lock file updates fail if declared README files are missing
# 3. Unlike pip/poetry, uv tools expect README files to be available
#
# Supports PEP 621 readme declarations:
# - readme = "filename.md" (string form)
# - readme = { file = "filename.md" } (table form)
# - Fallback to common filenames when not declared
#
# @return [Array<Dependabot::DependencyFile>] README files marked as support files
sig { returns(T::Array[Dependabot::DependencyFile]) }
def readme_files
```

**Comment by v-thavaahariharangit** [comment]
Consider adding additional test coverages

1. Explicit README Declaration Testing:
```
context "when pyproject.toml explicitly declares a README file" do
  let(:pyproject_with_readme) do
    {
      "project" => {
        "name" => "test-package",
        "readme" => "README.md"
      }
    }
  end

  before do
    # Stub pyproject.toml with explicit README declaration
    stub_request(:get, url + "pyproject.toml?ref=sha")
      .to_return(status: 200, body: fixture_with_content(pyproject_with_readme.to_toml))
    
    # Stub successful README fetch
    stub_request(:get, url + "README.md?ref=sha")
      .to_return(status: 200, body: github_file_fixture("README.md", "# Test Package"))
  end

  it "fetches the explicitly declared README as a support file" do
    files = file_fetcher_instance.files
    readme_files = files.select(&:support_file?)
    
    expect(readme_files.map(&:name)).to include("README.md")
    expect(files.find { |f| f.name == "README.md" }.support_file?).to be true
    expect(files.find { |f| f.name == "README.md" }.content).to include("# Test Package")
  end
end
```

2. Table Format README Declaration:
```
context "when pyproject.toml declares README in table format" do
  let(:pyproject_with_table_readme) do
    {
      "project" => {
        "name" => "test-package",
        "readme" => { "file" => "docs/README.rst" }
      }
    }
  end

  it "fetches the file specified in table format" do
    # Test that { file = "docs/README.rst" } format works
  end
end
```

3. Fallback README Discovery:

```
context "when pyproject.toml doesn't declare readme but common READMEs exist" do
  before do
    # pyproject.toml without readme declaration
    stub_request(:get, url + "pyproject.toml?ref=sha")
      .to_return(status: 200, body: fixture_without_readme)
    
    # First 3 README variants return 404
    stub_request(:get, url + "README.md?ref=sha").to_return(status: 404)
    stub_request(:get, url + "README.rst?ref=sha").to_return(status: 404)
    stub_request(:get, url + "README.txt?ref=sha").to_return(status: 404)
    
    # But plain README exists
    stub_request(:get, url + "README?ref=sha")
      .to_return(status: 200, body: github_file_fixture("README", "Plain text readme"))
  end

  it "falls back to discovering common README filenames" do
    files = file_fetcher_instance.files
    readme_files = files.select(&:support_file?)
    
    expect(readme_files.map(&:name)).to include("README")
  end
end
```

4. Error Handling Validation:
```
context "when declared README file doesn't exist" do
  before do
    # pyproject.toml declares README.md
    stub_pyproject_with_readme("README.md")
    
    # But README.md returns 404
    stub_request(:get, url + "README.md?ref=sha").to_return(status: 404)
  end

  it "gracefully handles missing declared README files" do
    expect { file_fetcher_instance.files }.not_to raise_error
    
    files = file_fetcher_instance.files
    readme_files = files.select(&:support_file?)
    expect(readme_files).to be_empty
  end
end
```

5. TOML Parsing Error Handling:
```
context "when pyproject.toml is malformed" do
  before do
    stub_request(:get, url + "pyproject.toml?ref=sha")
      .to_return(status: 200, body: '{"content": "' + Base64.encode64("invalid toml [[[") + '"}')
  end

  it "gracefully handles TOML parsing errors and still attempts README discovery" do
    # Should not crash, should fall back to common README discovery
  end
end
```

Why This is Critical
- Regression Prevention: Without these tests, future changes could break README fetching silently
- Functionality Proof: Current tests don't prove the feature actually works
- Edge Case Coverage: Real-world scenarios aren't tested
- Maintenance Safety: Changes to the logic could introduce bugs without detection

**Comment by v-thavaahariharangit** [comment]
Additional Documentations

1. In ADR
```
# UV FileFetcher README Support

## Decision
The UV FileFetcher will fetch README files referenced in pyproject.toml, 
unlike the main Python FileFetcher.

## Rationale
- uv and hatchling require README files during lock file generation
- Dependabot updates fail when declared README files are unavailable
- UV ecosystem has different requirements than pip/poetry

## Consequences
- Increases API calls per repository
- Creates behavioral difference from main Python ecosystem
- Solves real user issues with uv-managed projects

## Alternatives Considered
- Modify main Python FileFetcher (rejected: would affect all Python users)
- Conditional fetching based on build tools (rejected: too complex)
```

This is a nit pick, nice to have comment

**Review by thavaahariharangit — APPROVED** [review]
- **Clear Problem Statement:** Solves real issue where UV/hatchling needs README files during lock generation
- **Bounded Solution:** Limited to 4 common README filenames to prevent excessive API calls
- **Support File Pattern:** Correctly uses support_file flag for non-dependency files
- **Error Handling:** Gracefully handles 404s for missing README files

File: uv/spec/dependabot/uv/file_fetcher_spec.rb [hunk 4]
```diff
@@ -226,12 +226,160 @@
...
+            body: fixture("github", "contents_readme.json"),
+            headers: { "content-type" => "application/json" }
+          )
+      end
+
+      it "fetches the declared README as a support file" do
+        files = file_fetcher_instance.files
+        readme_files = files.select(&:support_file?)
+
+        expect(readme_files.map(&:name)).to include("README.md")
```
Comment (v-thavaahariharangit) [inline uv/spec/dependabot/uv/file_fetcher_spec.rb:269]: Purpose of the PR is tested here 👍
